### PR TITLE
Feature implib

### PR DESCRIPTION
**Pull request by brupelo**
Little improvement on top of my previous PR.

I had forgotten to test the implib feature when you specify the `--output-dir`, now it should work fine no matter whether you use `--output-dir` or not. Sorry I didn't test it properly before :/

### CODE REVIEW

**Comment by kayhayen** [comment]
Also kind of looks like you are not using "os.path.join()" but string "format", which is rather bad way of doing it. Never should one hard code "\\" not even in OS specific code, just because it makes it always harder to see what is going on, not to mention how it is better at handling corner cases.

As an aside, I also expect that Python 2.6 will not work, because "format" there only accepts keyword arguments, which is why I only use it like that in Nuitka. As you have probably noticed, I like named arguments very much. so I don't see that as an issue. Of course, using f strings of 3.6 would be more cool anyway, but that's not happening for a while.

But yeah, even those are worse than os.path.join. Honestly I don't even quite know what splitpath does on most days. :-)

Yours,
Kay

**Comment by kayhayen** [comment]
BTW: As for conflicts, do not bother, I will remove the merge and do another squash merge later.

**Comment by brupelo** [comment]
> Also kind of looks like you are not using "os.path.join()" but string "format", which is rather bad way of doing it. Never should one hard code "\" not even in OS specific code, just because it makes it always harder to see what is going on, not to mention how it is better at handling corner cases.
> 
> As an aside, I also expect that Python 2.6 will not work, because "format" there only accepts keyword arguments, which is why I only use it like that in Nuitka. As you have probably noticed, I like named arguments very much. so I don't see that as an issue. Of course, using f strings of 3.6 would be more cool anyway, but that's not happening for a while.
> 
> But yeah, even those are worse than os.path.join. Honestly I don't even quite know what splitpath does on most days. :-)

You're absolutely right here! Hardcoded paths like the ones I've provided in the commit should be avoided all the times, period. Thing why I've used in this case is mainly because in the past when I was writing a metacompiler that would produce visual studio commands (using cl & link) I would find myself than sometimes visual studio options (don't recall which ones specifically) weren't playing nicely when using `os.path.join` and I had to hardcode paths to produce the right outcome. In this case, I think `IMPLIB` should be fine with `os.path.join` so I'll definitely change it though :)

You know? These days I've almost forgotten about using `os.path` module at all... each time I see in some of my old codebase `os.path` stuff I try to replace it with the lovely `pathlib.Path` module... seriously, `pathlib.Path` is a wonderful module... I've falled in loved with that module the first time I've discovered it... `stem`, `with_suffix`, `with_name`, `suffix`, `name`, `path concatenation`, etc... The final code become much much less verbose and readable than when using the old `os.path` utilities :)

One fast question, which python version should I assume when improving nuitka? So far I've tried to avoid some stuff from python3.6, which it's the version I tend to use all the time... I almost don't use 2.6/2.7 anymore nowadays... and 3.3? I only use it when coding SublimeText plugins :/

> BTW: As for conflicts, do not bother, I will remove the merge and do another squash merge later.

That's great, but please, don't make it so easy to me hehe :) , I'd really like to eventually master git... so if there is any way I can squash the whole set of commits here and provide a PR with one single commit, let me practice and polish my poor git skills... Right now I was reading this interesting article talking about it [https://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit](https://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit)

**Comment by brupelo** [comment]
@kayhayen Hehe, ok, I've managed to make this PR into a single commit, how cool is that, eh? :'D . Please let me know if I've done something wrong here though, as I told you... I'm still learning how to become proficient with git :)

**Comment by kayhayen** [comment]
I am still confused, why you join the absolute and normalized source_dir in the first place, why not just source_dir. I have really bad experience with Scons and other tools on Windows with absolute paths, because many tools will happily build below strange paths, for as long as they do not appear on the command line. Then all sorts of errors pop up. Also it's not needed.

And then the basename of source_dir is not useful either, result_basepath would be better for that, but why not just something hard coded. I assume the name doesn't really matter, so why not use `os.path.join(source_dir, "import.lib")` and be done with it?

**Comment by brupelo** [comment]
Sure thing, if you consider the current commit could produce issues and if you also believe having hardcoded names for the import lib is just fine, please be my guest. Actually I can't think of any real-case where we'd like to preserve the original names of the import libraries... I mean, maybe if you wanted to use the dlls on some c++ projects the names could be relevant... but that's not very real case so I guess harcoded names that makes a simpler code is alright.

Btw, in cases like this one where you think some changes on top of it will produce a much cleaner/performant/elegant code, don't even ask me, just tweak it, merge it and I'll eventually test it on windows :) . Otherwise we'll be ending up reviewing milimetrically each little feature/PR hehe ;-D . Plus, mostly of the times I'll want to tweak anything I'll probably won't be able to choose the optimal way because I won't be aware about all the gritty details and implications.

I can give good advices at very high-level architectural decissions... but a very low-level I'm not sure I'll be very helpful (at least for now) ;)

PS. If something can be simplified, simplify it. If there is dead-code, get rid of it, and so and so forth... Less code means less maintenance. I'd used the normalize_path there as it's one functions I tend to use almost in all my projects, so..

\o

**Comment by kayhayen** [comment]
As for your suggestion that I could just do it myself. Indeed. It's a mistake a made in the past and it drives away contributors. As soon as somebody touched an area of Nuitka, I was also there and did improve it myself.The majority of people then consider themselves unnecessary or bad. So I have changed my stance, and actually will try and not follow impulses to just do it myself, and let other people know, why I think it should or must be different.

As for your "normalize", I dislike it, and I suggest you consider input sanitizing. Every value you get from the user side, must be checked, converted, then, and exactly then, when you got it. Not any later. Randomly normalizing paths in the middle of somewhere doesn't do any good, except bug hiding potentially. I would try and not do it. That is not to say that I have not done it wrong. But that is never the point. The code review is for 2 people to notice things and be rather sure of the outcome.

So yeah, I am not going to do this commit, even if it's fast. The whole point of this being a community is to get more stuff done as a whole, and to learn making better changes. If I get to do cleanups on your changes each time, that is not good.

And for your architectural skills, I am a skeptical thinker. The right time to believe something is when there is evidence for it. Somebody telling me he has skills, doesn't count as evidence. Somebody actually applying them successfully does.

The idea of your other PR is very interesting. It feels like it should be a best practice actually. I am pretty sure I have not seen it in Python projects I fancy. I have applied "property sheets" (that is the name I know for this) for GUIs, where these values are also dynamic. I can totally see the benefit for a CLI now having seen your example, that is great.

But as you can imagine, I find this innovative and cool, and even very useful. I am not so convinced this is really a break through that I must be working on, now that I have the idea. Unlike what you said elsewhere, I am not spending a lot of time upfront on infrastructure. In fact, when I set out to prototype the initial design of Nuitka, it was a feasibility study only, with the purpose of demonstrating that it can be done at all and achieve real work.

Compiling huge Python programs like Mercurial correctly was a goal. Initially I set out and used C++11 and Python code templates as well as C++ template functions and classes on the backend side, and was working from bytecode on the frontend side. I had to give up on bytecode, because there was so much that it could never yield information about. And I had to give up on C++ mainly for the exception handling not performing at all, something I didn't anticipate. 

I also had an sad optimization phases that was only indicating the possiblity that I will introduce proper SSA later, but really only could do peephole stuff, which since then I have replaced with that, first variable traces, more recently also value tracing.

I am normally relatvely good at leaving open spots for things, from where I can generalize, e.g. having a "CType" class for the "PyObject *" which got many friends immediately, and more interesting ones recently.

I do not mind, that only when doing it for "bool", I realized which way the API should evolve, that that actually the accessed value is its own type, and needs richer interface, than mere storage classes, i.e. "PyCellObject *" grants access to a "PyObject *" value, where more operations will happen, than ever on cells. So far, these had be all joined, but on factory it's more obvious.

The API of that, is now very ugly. I did cleanups of it, but more will be needed. In fact, there should be two different kinds of classes, and I think there is still base classes with code that is only used for one child, it's messy. But insanely powerful results. Yeah, but cleanups to do.

The issue with Nuitka and design mainly for me is that the requirements were and are largely unknown to me, until I actually do it. My goto generators approach worked out as originally designed, and that felt cool for once, but the Ctype thing is a total unknown to me. But rather I know it will evolve further if I go from "bool" (complete for 0.6.0) via "void"  (should be complete already, but enabling will happen 0.6.1 likely) to "int", not sure how long that will take.

I really think Nuitka, unlike other software that I have designed, is a prototype project that gradually turns more and more into the real thing. I have literally spent years to inject proper design in steps into optimization phase, what I call SSA, value tracing, and it is very much there now. I am probably going to spend similar amounts of time, to execute on applying type inference results to the code generation.

So I turned that into something working with strings to something working with variable declaration objects knowing their type for the goto generators, aiming at C types. All the while carrying the full weight of passing every test here is.

Then e.g. cleaning up module variables to no longer have their special branch, but a pseudo C type, that makes them like everything else. Great. But when I first introduced the new thing, I postponed that, because I could sooner apply its benefits to some things and get experience from it.

Making a whole Nuitka design upfront and then executing it, would instead give a very high probability of failing in the real world. I am therefore applying the more agile approach, where I make things work. And then continue to work while I clean it up. For every feature I added, I actively go out, and change the thing, that made it hard. Always. I think Nuitka is largely developed by cleanups and refactoring. Goto generators were a fine  example of that, solving many of the issues by injecting variable declarations objects into code generation, made it easy to indicate storage (heap or object or stack) right there. 

That is not to say that Nuitka didn*t have the typical compiler design. Like parsing inputs, optimizing a tree internally, producing outputs. But that grand top level design only tells you the obvious things really and is stolen anyway from knowing similar projects like gcc.

I guess, I am an extreme programmer, or agile, or however they call it these days. At least for Nuitka. In my professional life, I have designed software for ATC on the drawing board, then in paper, and then in code, the design just worked, and got operational right after completion, which is rare I can tell you.

But the complexity of Nuitka is so dramatically higher. it is doing a complicated, i.e. detail rich work, and then it also is doing hard jobs where many things play together. And the wish to have something working before it is completed, if it ever is, makes things very different from projects I typically did. So the first version of Nuitka already had a use, and the desire was to evolve gradually.

I think I have desribed this elsewhere, but for large parts of the well or bad designed solutions of Nuitka, there is reliable ways of demonstrating it works. Far better than I have ever encountered. i believe it's the main reason I managed to get this off the ground is that. Having a test "oracle" is what makes Nuitka special, i.e. comparing to existing implementations. Like a calculator can be tested comparing it to one of the many already perfect ones out there. 

So for me, Nuitka is on the design level, something that goes through iterations, discovery, prototyping, and is actually really exciting in that. Compilers typically are boring. But for Nuitka that is totally not the case, because Python is not made for it. Well, that*s technically untrue, lets say not for optimizing compilers, not for type inference, etc.

Coming back to you, and some of that content above is for TWN 6 that I am supposed to write today, you need to decide if you want to finish what you start or not. I think you are on a nice path of getting into it. But I have other PRs, where I tell people my remarks, and if they agree, they don't make a lot of arguments, they just push the fixup commit, and I merge, esp. when it's on this level of complexity.

Also I did mention that `"{}".format("something")` will not work on Python 2.6, which I am using it very little in Nuitka, and if I do, always like this: `"{name}",format(name = "something")` as that is the only way it works there too. And no, I am not interested in explaining why 2.6 is relevant still. I prefer the notation `"%(name)s" % { "name" : "something" }` for its flexibility, or if it's short and doesn't benefit from names, like it is the case here, or wait, it is not going to be needed at all anymore, right.

I guess I should add those things to the coding rules I guess.

**Comment by brupelo** [comment]
> As for your suggestion that I could just do it myself. Indeed. It's a mistake a made in the past and it drives away contributors. As soon as somebody touched an area of Nuitka, I was also there and did improve it myself.The majority of people then consider themselves unnecessary or bad. So I have changed my stance, and actually will try and not follow impulses to just do it myself, and let other people know, why I think it should or must be different.

That's really interesting and a great lesson to learn from. I understand what you mean here and that feeling, actually I think you're absolutely right with that position. 

The fact you've thought about it speaks very nice from you, (ie: you've identified a problem that happened to you before and you have taken measurements to try to fix it and prevent it to happen again). Pointing out this cos many people don't have the slight bit of self-criticism and when they've made something "wrong" they won't try to fix it nor thinking about it not even a bit, IMHO that's really bad attitude.

For a project to become succesfull eventually you just don't need talented coders all around, you need those coders to become good mentors for the newcomers to grow up, talented coders that get people motivated.... that's the ideal.

On the other hand, while it's true in OSS there isn't anything like "managers/seniors/..." I do believe repo harcoders/authors should take this sort of responsability a little bit more serious sometimes... at least with large projects. To handle large projects you need team splitting and delegation, period.

> That is not to say that I have not done it wrong. But that is never the point. The code review is for 2 people to notice things and be rather sure of the outcome.
> 
> So yeah, I am not going to do this commit, even if it's fast. The whole point of this being a community is to get more stuff done as a whole, and to learn making better changes. If I get to do cleanups on your changes each time, that is not good.

I understand and makes sense. As I told you, my git skills are really poor right now and my contribution to Open Source projects is mostly of the times about trivial stuff (don't involve really complex git workflows). In this case, nuitka is a project that has catched my attention very much and really interests me, as I've told you zillion of times I think the project itself will become eventually a really fast/stable tool people can rely on when it comes to create standalone python deployments. So improving my skills as a contributor here definitely interests me. So yeah, thanks to point this out! I do think your stance here is the most benefitial one in the long term.

> And for your architectural skills, I am a skeptical thinker. The right time to believe something is when there is evidence for it. Somebody telling me he has skills, doesn't count as evidence. 

Just for the record, don't get me wrong, I'm not claiming I've got architectural skills but it's more like at this moment I don't have the low-level knowledge in my head about the code to make the best decissions mostly of the time. In order to make the best decissions you really need to be aware about the existing code, so you'll avoid code duplication or hacks. Ask yourself, why newcomers add project hacks all the time when get involved on new projects? That's mainly because they don't know the project but they still need to get the job done... yet another proof than tight deadlines go in detriment of the project quality. And of course, to reach that level of awareness you need to become really familiar with the code/framework itself, which in this particular case, it's not trivial as the project is already medium/large.

Btw, I think I've already explained this in other threads but let me try again, just to be sure we're on the same page. If sometimes I sound like a "smarty that knows everything" ass please take my words with a little grain of salt, period. I'm not the type of person who likes to prove anything to anybody... I just really love coding and I love teaching from experience... but the most important, I do love to be teached and learning new stuff about almost everything.

So yeah, when I talk about stuff or giving some tips about how to improve something is just based on succesfully personal experience, that's all... you'll never hear from me the solutions I propose are the only&right way to go and actually, I do like to be proved wrong when necessary, that's the best way to learn. You don't teach a child by not correcting his bad behaviour, right? Same with us, professional developers, I think it'd be really wrong to assume we know almost everything and our ways are the only ways... out there you'll see coders with a lot of ego but I think that's just wrong. Does it make sense?

Actually, I'm quite a self-critical person and while I can't deny I've got quite the confidence with my skills mainly because I've coded since I was a little kid for more than 20 years on almost on any type of language and created almost any type of project my ego is low :)

> Somebody actually applying them successfully does.

Well, you know? In an ideal world I would be able to spend all my time/efforts on hobby projects as interesting as nuitka but that's not the case and spare time is really limiting, so not sure how many evidences you'll be able to see here haha :)

If you check my github profile you'll see I haven't uploaded any of my personal projects and mostly of my contributions to OSS have been really lightweight/trivial, I have never got involved in a serious way on any github existing project tbh... and that doesn't mean I don't believe in open source philosophy but you won't see me uploading any of my private code anytime soon, why? Because OSS takes quite a lot of time and effort :/ . Dealing/interacting with people, delivering features to make people happy, answering casual users, answering casual devs, answering casual question, all of it really takes time. And don't get me wrong, I think all of it is fun & rewarding but in my case I'm dealing with dozens of active projects in a daily basis so. 

For instance, take a look below how I store source code, in the others folders I keep projects from companies/forks/github/others...  in the personal one I keep all the stuff I've coded... And no, that size isn't because databases/assets/3rdparty libraries/static_files ... :/ . When I say I'm dealing with dozens of projects in a daily basis I really mean it ;D

personal

> The idea of your other PR is very interesting. It feels like it should be a best practice actually. I am pretty sure I have not seen it in Python projects I fancy. I have applied "property sheets" (that is the name I know for this) for GUIs, where these values are also dynamic. I can totally see the benefit for a CLI now having seen your example, that is great.

Kind of unrelated but you know what? I do really like Eclipse plugin framework... I've adopted some of the concepts to some of my plugin frameworks, the only bad about eclipse is the iteration is slow because the nature of the language but if you apply the same philosophy with interpreted languages like python you end up doing it the right way. 

Usually I like the approach where you can run your software 0 plugins or with hundred of them. In fact, the "everything is a plugin" is great, you know why? It allows team splitting! not sure how much you're into listening podcasts but if sometime you feel like it, here's a really nice [podcast](http://www.se-radio.net/2008/07/episode-104-plugin-architectures/) to listen to. I've listened few times and it exposes a lot of interesting ideas/concepts you can learn from.

> I guess I should add those things to the coding rules I guess.

Actually, that's something I've asked and I think you missed on previous threads, which python version coding style should we consider to nuitka?   2.6, 2.7, ...?

Anyway, sorry for not commenting upon your other stuff but I need to leave now, plus I guess that's content you'll add into your blog anyway so it wasn't directed specifically to me. Thanks for such a detailed explanation from yours anyway, I love those really big chunks of text :) . Nowadays people tend to write in a very twitter-way for almost everything... and while sometimes I understand the point of it (writing is hard and takes effort) I really dislike the twitter-way to communicate with people hahah and you'll see very twits from me :D . Hopefully my writing is not very wrong and boring to read, but yeah, improving my English writing skills is already on my TODO list.

Yours,
Bruno.

Ps. Oh, yeah, about your review, today I'll try to find the time to update this PR. While I was writing this I could have updated already, couldn't have? lol :D

**Comment by brupelo** [comment]
One last question, even if it doesn't matter too much as I've already clarified on my previous post. Why do you think in this particular is better using a harcoded name for the import lib instead a more meaningful one?

I mean, for this particular case I don't see you're losing any type of information... if you decided to grep all the importlibs of your build directories you'd get the name of the build folder themselves but, answer me something. If a solution is simpler at expenses of not storing "not-useful-but-maybe-relevant-information-in-the-future-or-not" information, would be that a good thing?

Asking you cos sometimes I struggle about that... ie: Should i leave space to store metadata that can be used in the future (or not), should i add extra fields on this data structure for future usage (or not)?

Anyway, this PR **should be ready to be merged**, in case you find anything else to tweak please let me know y I'll update.

B.

**Comment by kayhayen** [comment]
Going to merge it the other day. I am very close to making a new release, and don't want to mess with that. But this make it for 0.6.0 and next pre-release surely.

Thanks for your contribution.

As for your question, I think "DRY" applies. Do not repeat yourself unless necessary. With ``dirname/dirname.ext`` and that file being considered for use outside of that directory ever, it would make most sense.

Other example is ``reponame/.git`` where that is not going to happen, and then you want to label it as, yeah, this is the git config for that specific folder.

In my view, the build directory is much like an inseparable something. And that .LIB and .EXT file are garbage to being with anyway. I don't see much value in keeping it at all after the linker has succeeded.

**Comment by brupelo** [comment]
Cool, all clear.

> As for your "normalize", I dislike it, and I suggest you consider input sanitizing. Every value you get from the user side, must be checked, converted, then, and exactly then, when you got it. Not any later. Randomly normalizing paths in the middle of somewhere doesn't do any good, except bug hiding potentially. I would try and not do it. That is not to say that I have not done it wrong. But that is never the point. 

I think you're right here, input parsing/sanitizing/checking/conversion should be done by the first modules/plugins/subsystems so they'll provide sanitized data to the subsequent subsystems. In fact, sometimes my transformation functions tend to accept very random input data (they have weak preconditions) so they end up just doing too much, that's not good :/ . This refers to what I've explained on some previous post about newcomers providing hacks to projects they start contributing to... they want to add a particular feature but they don't want to dig further low-level to figure out what's the proper way to do it... At the end of the day, everybody like high-level abstractions without know whether there is other type of code already achieving the task, in this particular case I didn't want to read through the code to know what was exactly the meaning of those path variables used on the scons file.

Btw, let me tell you a little secret I always like to apply when coding my stuff :) ... When I code any kind of module the first thing I like to consider is this, "alright, If I had to pay some coders to improve/extend/use my stuff and they were going to charge me per hour, what'd be the cheaper way to do so....?" You know what's the answer? actually it is an answer every coder learn almost when starting to code, proper level of abstraction of code.

If there is a system composed by {A,B,C,D,E,F} and C->D and I wanted somebody else to improve C I'll just want him to know about C public interfaces... and if I want him to improve D I'll just want him to know about C&D. But definitely I've got no interest they know not a single bit about A,B,E,F (not even the public interfaces). I'll keep them like black-boxes

Alright, that's all I wanted to say, not posting anything else in this thread unless the feature itself needs more changes.

> Thanks for your contribution.

No, thanks to you to listen to my non-required long and probably boring "speeches" ;D . I'll eventually would like to write a book/blog but haven't found yet the time to do so :)

**Comment by kayhayen** [comment]
@brupelo Writing a book is on my list too, so we have that in common. :)

As far as blackbox goes, the scons part is pretty well interfaced part of Nuitka with a really clear interface from our side.

And my replacement for that cost evaluation is laziness and necessity. I think a coder must be lazy, but do what it necessary, then he will automatically become good.

**Comment by kayhayen** [comment]
@brupelo Now finish the "wip" and I would love to add you for that to the contributors list. :)

**Comment by kayhayen** [comment]
I did a small amend for this one, so Python 2.6 works (I take that on my all the time), and removed the now unused normalize_path function. I hope it doesn't break from me doing it, but I am glad it will then have your name attached. ;-)

**Comment by brupelo** [comment]
> @brupelo Writing a book is on my list too, so we have that in common. :)

Cool, when you publish let me know and if the subject is something I'm interested in maybe I'll buy you 1 ;)

> @brupelo Now finish the "wip" and I would love to add you for that to the contributors list. :)

Contributor list? Oh... I thought I was already added on the "most chatty contributors list" hehe ;'-)

One little aside note about me, so you know me a little bit, if some project catches my interest I've got this "habit" about becoming quite passionate about it, I guess that's pretty obvious after all text chunks I've been posting here at github the last days. Usually I don't have "middle term", either I'm quite passionate about a project or I just lose completely interest about it. 

Usually I'm the type of person who likes to reinvent the wheel all the time and I like my custom solutions before considering existing tool. But if there is some piece of tech I don't fully understand that intrigues me quite a lot... and nuitka, a one-man project such as nuitka inspires me tbh. I respect people who thinks out of the box and I respect that, as I usually do that as well, so yeah, hats off.

On the other hand, I wouldn't call myself a very good team player, that's something I've realized few years ago, the good thing about that realization is thanks I've been aware about it I've tried to put quite a lot of effort to improve that "skill"... I think while being an "I can do it all by myself" type of person is quite rewarding in the long term it proves to pretty limiting and slow.

About the wip branch... hehe, yeah, when I find out the spare time budget I'll definitely come back to it, no worries ;) . One of the main reasons about that branch was I wanted to use my own clis on top of nuitka so I wouldn't have to bother you with very specific issues that maybe are interesting for some of my use-cases. 

Right now one of my local branchs contains some custom options that modifies MainControl behaviour to do some custom stuff and it's getting me the job done... But of course, that's not feasible in the long term and I'd like eventually having the wip branch finished.

**Comment by kayhayen** [comment]
@brupelo One thing I meant to get back on, is how you said:

> The fact you've thought about it speaks very nice from you, (ie: you've identified a problem that
> happened to you before and you have taken measurements to try to fix it and prevent it to happen 
> again). 

That is how I tend to think of me myself. I found it very nice to see that you got that. I cannot reply to all your context, but I will most often read it.

I am abusing this PR for meta communication now. I have no reason to know, but when you say, you might not stay in a project, I feel that your switch is likely more often caused by you getting on the nerves and negative feedback.

Therefore I want to be clear. I like energy, it makes for creative stuff. I even more like action derived from ideas. If you are going to do that, all fine.

I noticed one thing, and that is when 3rd parties are involved, e.g. a change suggestion that you disagree with, I am having a hard time, to want and let you speak with authority there. I read that you said, people should never take you serious. I know that attitude. It's easy and convenient, but also irresponsible. Others do not know that, and nor should they be required to.

As you noticed, flat out criticism, on spot or wrong, both don't irate me at all. However I know for a fact that people really and rightfully do not appreciate to get questions and pre-emptive answers along.

I think the right approach after asking a question in that issue, where I asked for deletion, was to wait for answers. And stating facts is OK, but making personal conclusions I don't think so.

OK: Why do you want to have this feature, potentially the solution may not satisfy you, so we are interested in known why you want this.

OK: When solved this will produce very large binaries and then it's having problems when delivering multiple binaries still. Do you do this?

I am stressed, because to a reader, it might simply discourage if you merely express disdain, because if you are going to the length of making a PR, you need to 100% know, or want to know, that it is not going to rejected, because e.g. I share your sentiments. Therefore I am forced to reject them and make it clear, and in best case, I am not only loosing time on that, but more important what would actually be a useful contribution and potentially new contributor for Nuitka to gain even more traction.

So if you go there, I am not happy with you doing it. This is never about not caring what you think. In fact I do think. For that issue however, I found your comments to be irrelevant. Occasionally that's bound to happen, isn't it?

Factually you are also incorrect for the worry about larger binaries. People deliver whole docker images these days, for trivial things. These might be well larger. Surely the ".dist" folder derived can only be smaller. And even then, people will be free to not care.

In fact, occasionally I hear complaints about ".dist" size, but rarely so. Most people seem OK, and not everybody is using PyQt, some people use this for really small stuff. The "clcache" binary for e.g. would be nice candidate.

And one thing is absolutely clear. Forever, people are going to think of compilers as capable of producing running binaries that are single files. They just are. Forever these are going to find directories cumbersome, that is bound to happen. If a subset of these people solves it eventually, and standalone was much similar, I didn't do it myself, I will gladly carry it.

But going back, I would really want to delete our comments there. They are noise. If that offends you, I am not doing it, that is why I asked you if it's OK. But I am surely unhappy and ask you to be more sure to be constructive in sharing opinions about enhancements that you don't care about.

Also I would very much like it, it actually did and went out and tried to represent the project, like with new issues as you already do. That makes me happy.

In my mind, there is only little coaching to be done. If you think this is impossible for you, I would disagree. But maybe you don't like to constrain yourself in public interactions. I would understand, you are not obliged to anything.

I occasionally try to change. I recommend trying it. :-)
